ufl/algorithms/apply_derivatives.py

```diff
@@ -165,25 +188,65 @@ def independent_operator(self, o):
         """A zero with correct shape and indices for operators independent of diff. variable."""
         return Zero(o.ufl_shape + self._var_shape, o.ufl_free_indices, o.ufl_index_dimensions)
 
+    # --- Error checking for missing handlers and unexpected types
+
+    @singledispatchmethod
+    def process(self, o: Expr) -> Expr:
+        """Process ``o``.
+
+        Args:
+            o: `Expr` to be processed.
+
+        Returns:
+            Processed object.
+
+        """
+        raise AssertionError(f"UFL expression expected: got {o}")
```

Comment:
Is this now boilerplate that needs reproducing for each subclass? Might it be better on the ABC?

Reply:
Do you mean putting this base case in the base class? I think the problem is that `singledispatchmethod` object needs to be made in each child class. What I had to do here is probably relevant https://github.com/firedrakeproject/ufl/blob/c16f038378e17ad8bd730bf9a020028ad72a00c1/ufl/algorithms/apply_derivatives.py#L716.

Reply:
maybe call super like you do in the other case.

Reply:
Done.